Add the ability to opt into logging on a per-endpoint basis

bradykieffer opened the pull request:
Also, expose conditionals that will by default opt all endpoints into
being logged.

Note one change, responses can now be logged without an actual request
being logged.

This may be a little more opinionated than needed but I personally prefer _not_ using an attribute on a view method as a flag on whether or not to log the method. For newer versions of Python dictionaries are fairly compact so I think storing methods that we don't want logged / want to opt into logging is a reasonable thing to do (and I think it makes the code a lot cleaner)

bradykieffer commented on test_urls.py:
```diff
@@ -2,8 +2,9 @@
+from django.http import request
 from django.views import View
-from request_logging.decorators import no_logging
+from request_logging.decorators import no_logging, opt_into_logging
```
I need to add a testcase for this

tclancy commented on request_logging/middleware.py:
```diff
@@ -107,14 +132,14 @@ def __call__(self, request):
-            if because is not None:
-                return self._skip_logging_request(request, because)
+        should_log_route = self._should_log_route(request)
+        if not should_log_route.log_route:
```
I realize this mirrors the old code, but why the nested `if`s instead of one `and`? Also, it seems like we should dump out in either case so it should be an `or` (and maybe we provide an empty string if there's no `skip_reason`).

tclancy commented on request_logging/middleware.py:
```diff
@@ -184,28 +218,35 @@ def _log_request_body(self, request, logging_context):
+        if should_log_response:
+            # Either the response is opted-in to logging by default or we've
+            # conditionally selected the response to be logged. Regardless, log it!
+            if 400 <= response.status_code < 500:
```
Please break this out into its own function,

bradykieffer commented:
@tclancy apologies for this taking so long to get back to, I've been in the middle of a move and unable to get to this. I _think_ I've addressed your feedback though (and hoping that tests all pass of course)

bradykieffer commented on request_logging/middleware.py:
```diff
@@ -141,9 +191,19 @@ def _should_log_route(self, request):
+
+        return func
+
+    def _should_log_route(self, request):
+        func = self._get_api_func(request)
+        if func in OPT_INTO_LOGGING_FUNCS:
+            return ShouldLogRoute(log_route=True, skip_reason=None, silent=False)
+
+        if func in NO_LOGGING_FUNCS:
+            no_logging = NO_LOGGING_FUNCS[func]
+            return ShouldLogRoute(log_route=False, skip_reason=no_logging.message, silent=no_logging.silent)
+
+        return ShouldLogRoute(log_route=self.logging_opt_in_defaults[REQUEST](request), skip_reason=None, silent=False)
```
I couldn't really come up with a better way to handle the silent case than propagating a flag forward. It definitely feels a little leaky but it works with current behavior

bradykieffer commented:
Closing this because it's super old and I don't have the time to update this 😢
